pkg/controller/performanceprofile/performanceprofile_controller.go

```diff
@@ -103,54 +208,225 @@ func (r *ReconcilePerformanceProfile) Reconcile(request reconcile.Request) (reco
 		return reconcile.Result{}, err
 	}
 
-	// Define a new Pod object
-	pod := newPodForCR(instance)
+	if instance.DeletionTimestamp != nil {
+		// delete components
+		if err := r.deleteComponents(instance); err != nil {
+			klog.Errorf("failed to delete components: %v", err)
+			return reconcile.Result{}, err
+		}
 
-	// Set PerformanceProfile instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, pod, r.scheme); err != nil {
-		return reconcile.Result{}, err
+		// remove finalizer
+		if hasFinalizer(instance, finalizer) {
+			removeFinalizer(instance, finalizer)
+			if err := r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// TODO: we need to check if all under performance profiles values != nil
+	// first we need to decide if each of values required and we should move the check into validation webhook
+	// for now let's assume that all parameters needed for assets scrips are required
+	if err := r.verifyPerformanceProfileParameters(instance); err != nil {
+		// we do not want to reconcile again in case of error, because a user will need to update the PerformanceProfile anyway
+		klog.Errorf("failed to reconcile: %v", err)
+		return reconcile.Result{}, nil
 	}
 
-	// Check if this Pod already exists
-	found := &corev1.Pod{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
-		if err != nil {
+	// add finalizer
+	if !hasFinalizer(instance, finalizer) {
+		instance.Finalizers = append(instance.Finalizers, finalizer)
```

Comment:
To my best knowledge it's better to update a copy of the instance, for the same reason David mentioned at another place (updating a chached instance from informer is bad)?

Reply:
I unsure that we modify cached instance, we create a new one and fill it with the details from the cached instance
```
instance := &performancev1alpha1.PerformanceProfile{}
err := r.client.Get(context.TODO(), request.NamespacedName, instance)
```

I believe the main reason to copy, it when you want to change something under the instance for the update, but still want to continue reconcile loop and work on the instance without update

Reply:
> we create a new one
oh yes, sure, sorry, missed that